src/flood_fill.jl

```diff
@@ -1,35 +1,139 @@
-flood_fill(f, src::AbstractArray, idx::Union{Integer,CartesianIndex}, nbrhood_function=diamond_iterator(window_neighbors(src))) =
+"""
+    mask = flood(f, src, idx, nbrhood_function=diamond_iterator((3,3,...)))
+
+Return an array `mask` with the same axes as `src`, marked `true` for all elements of `src` that:
+
+- satisfy `f(src[i]) == true` and
+- are connected by such elements to the starting point `idx` (an integer index or `CartesianIndex`).
+
+This throws an error if `f` evaluates as `false` for the starting value `src[idx]`.
+The sense of connectivity is defined by `nbrhood_function`, with two choices being
+[`ImageSegmentation.diamond_iterator`](@ref) and [`ImageSegmentation.box_iterator`](@ref.)
+
+# Examples
+
+```jldoctest; setup=:(using ImageSegmentation, ImageCore)
+julia> mostly_red(c) = red(c) > green(c) && red(c) > blue(c)
+mostly_red (generic function with 1 method)
+
+julia> img = repeat(LinRange(colorant"red", colorant"blue", 4), 1, 2) # red-to-blue
+4×2 Array{RGB{Float32},2} with eltype RGB{Float32}:
+ RGB{Float32}(1.0,0.0,0.0)            RGB{Float32}(1.0,0.0,0.0)
+ RGB{Float32}(0.666667,0.0,0.333333)  RGB{Float32}(0.666667,0.0,0.333333)
+ RGB{Float32}(0.333333,0.0,0.666667)  RGB{Float32}(0.333333,0.0,0.666667)
+ RGB{Float32}(0.0,0.0,1.0)            RGB{Float32}(0.0,0.0,1.0)
+
+julia> flood(mostly_red, [img; img], 1)   # only first copy of `img` is connected
+8×2 BitMatrix:
+ 1  1
+ 1  1
+ 0  0
+ 0  0
+ 0  0
+ 0  0
+ 0  0
+ 0  0
+```
+
+See also [`flood_fill!`](@ref).
+"""
+flood(f, src::AbstractArray, idx::Union{Integer,CartesianIndex}, nbrhood_function=diamond_iterator(window_neighbors(src))) =
     flood_fill!(f, falses(axes(src)) #=fill!(similar(src, Bool), false)=#, src, idx, nbrhood_function)
 
-function flood_fill(src::AbstractArray, idx::Union{Integer,CartesianIndex},
+function flood(src::AbstractArray, idx::Union{Integer,CartesianIndex},
                     nbrhood_function=diamond_iterator(window_neighbors(src)); thresh)
     validx = src[idx]
     validx = accum_type(typeof(validx))(validx)
     return let validx=validx
-        flood_fill(val -> default_diff_fn(val, validx) < thresh, src, idx, nbrhood_function)
+        flood(val -> default_diff_fn(val, validx) < thresh, src, idx, nbrhood_function)
     end
 end
 
-function flood_fill!(f, dest, src::AbstractArray, idx::Union{Int,CartesianIndex}, nbrhood_function=diamond_iterator(window_neighbors(src)))
+"""
+    flood_fill!(f, dest, src, idx, nbrhood_function=diamond_iterator((3,3,...)); fillval=true, isfilled = isequal(fillval))
```

Comment:
For consistency to [ImageTransformations](https://juliaimages.org/ImageTransformations.jl/dev/reference/#ImageTransformations.imrotate) where I use `fillvalue` as the keyword name, maybe rename to `fillvalue`?